packages/common/src/store/pages/chat/sagas.ts

```diff
@@ -67,16 +69,46 @@ function* doFetchMoreMessages(action: ReturnType<typeof fetchMoreMessages>) {
   try {
     const audiusSdk = yield* getContext('audiusSdk')
     const sdk = yield* call(audiusSdk)
+
+    // Ensure we get a chat so we can check the unread count
+    yield* call(fetchChatIfNecessary, { chatId })
+    const chat = yield* select((state) => getChat(state, chatId))
     const summary = yield* select((state) =>
       getChatMessagesSummary(state, chatId)
     )
-    const before = summary?.prev_cursor
-    const response = yield* call([sdk.chats, sdk.chats!.getMessages], {
-      chatId,
-      before,
-      limit: 15
-    })
-    yield* put(fetchMoreMessagesSucceeded({ chatId, response }))
+
+    // Paginate through messages until we get to the unread indicator
+    let lastResponse: TypedCommsResponse<ChatMessage[]> | undefined
+    let before = summary?.prev_cursor
+    let hasMoreUnread = true
+    let data: ChatMessage[] = []
+    while (hasMoreUnread) {
+      const limit = 50
+      const response = yield* call([sdk.chats, sdk.chats!.getMessages], {
+        chatId,
+        before,
+        limit
+      })
+      // Only save the last response summary. Pagination is one-way
+      lastResponse = response
+      data = data.concat(response.data)
+      hasMoreUnread =
```

Comment:
can't say i'm totally following this logic yet but I trust!

Reply:
basically fetch until the number of newer messages is > the number of unreads we expect

Reply:
added comment and checked my math 🤔